pkg/controller/kfservice/resources/istio/virtualservice.go

```diff
@@ -0,0 +1,230 @@
+}
+
+func createFailedStatus(reason string, message string) *v1alpha2.VirtualServiceStatus {
+	return &v1alpha2.VirtualServiceStatus{
+		Status: duckv1beta1.Status{
+			Conditions: duckv1beta1.Conditions{{
+				Type:    v1alpha2.RoutesReady,
+				Status:  corev1.ConditionFalse,
+				Reason:  reason,
+				Message: message,
+			}},
+		},
+	}
+}
+
+func (r *VirtualServiceBuilder) CreateVirtualService(kfsvc *v1alpha2.KFService) (*istiov1alpha3.VirtualService, *v1alpha2.VirtualServiceStatus) {
+
+	httpRoutes := []istiov1alpha3.HTTPRoute{}
+
+	// TODO (rakelkar) get these from config
+	ingressGateway := "knative-ingress-gateway.knative-serving"
+	destinationServiceName := "istio-ingressgateway.istio-system.svc.cluster.local"
+
+	// destination for the default predict is required
+	predictDefaultSpec, defaultPredictReason := getPredictStatusConfigurationSpec(&kfsvc.Spec.Default, kfsvc.Status.Default)
+	if predictDefaultSpec == nil {
+		return nil, createFailedStatus(defaultPredictReason, "Failed to reconcile default predictor")
+	}
+
+	// extract the virtual service hostname from the predictor hostname
+	serviceHostname := extractVirtualServiceHostname(kfsvc.Name, predictDefaultSpec.Hostname)
+	serviceURL := fmt.Sprintf("http://%s/v1/models/%s", serviceHostname, kfsvc.Name)
+
+	// add the default route
+	defaultWeight := 100 - kfsvc.Spec.CanaryTrafficPercent
+	canaryWeight := kfsvc.Spec.CanaryTrafficPercent
+	predictRouteDestinations := []istiov1alpha3.HTTPRouteDestination{
+		createHTTPRouteDestination(predictDefaultSpec.Hostname, defaultWeight, destinationServiceName),
+	}
+
+	// optionally get a destination for canary predict
+	if kfsvc.Spec.Canary != nil {
+		predictCanarySpec, canaryPredictReason := getPredictStatusConfigurationSpec(kfsvc.Spec.Canary, kfsvc.Status.Canary)
+		if predictCanarySpec != nil {
+			canaryRouteDestination := createHTTPRouteDestination(predictCanarySpec.Hostname, canaryWeight, destinationServiceName)
+			predictRouteDestinations = append(predictRouteDestinations, canaryRouteDestination)
+		} else {
+			return nil, createFailedStatus(canaryPredictReason, "Failed to reconcile canary predictor")
+		}
+	}
+
+	// prepare the predict route
+	predictRoute := istiov1alpha3.HTTPRoute{
+		Match: []istiov1alpha3.HTTPMatchRequest{
+			istiov1alpha3.HTTPMatchRequest{
+				URI: &istiov1alpha1.StringMatch{
+					Prefix: fmt.Sprintf("/v1/models/%s:predict", kfsvc.Name),
+				},
+			},
+		},
+		Route: predictRouteDestinations,
+	}
+	httpRoutes = append(httpRoutes, predictRoute)
+
+	// optionally add the explain route
+	explainRouteDestinations := []istiov1alpha3.HTTPRouteDestination{}
+	if kfsvc.Spec.Default.Explainer != nil {
+		explainDefaultSpec, defaultExplainerReason := getExplainStatusConfigurationSpec(&kfsvc.Spec.Default, kfsvc.Status.Default)
+		if explainDefaultSpec != nil {
+			routeDefaultDestination := createHTTPRouteDestination(explainDefaultSpec.Hostname, defaultWeight, destinationServiceName)
```

Comment:
what if there is 20% canary traffic specified, but no explainer on the canary spec, only 20% traffic goes to default?

Reply:
I though we decided that shape cant be changed between default and canary?

Reply:
looks like we still have not reached a consensus(https://github.com/kubeflow/kfserving/issues/324) :), I am ok to keep this assumption for now.